### PR TITLE
create-config CLI and using config file for beacon CLI

### PR DESCRIPTION
**Pull request by austinabell**
TLDR: Implemented #177, refactored/removed unnecessary imports/added typing for files I came across or directly relating, added tests including testing the required output file flag (assuming functionality of throwing error when -o flag not set which was implied)

I implemented the TOML marshalling and unmarshalling using the library:

```
"@iarna/toml": "^2.2.3",
```

Because it was the only one that had the stringifying and parsing functionalities needed and seemed well tested.

I had an issue with the deepmerge library that was used in the project for my specific case. The problem I had was if an empty object (which had a parameter defined and defaulted to be undefined) was deep merged with an object that had that type defined, it would not be merged and instead kept the empty object value.

for example:

```Typescript
db: {
    name: options.db
},
```

would create (when options.db was undefined):

```
"db":{}
```

and when merged with:
```
"db":{"name":"..."}
```
would end up being:
```
"db":{}
```

And when trying to keep the heirarchy of `cli options > config file options > defaults` there would be inconsistencies and unintended functionality. Workaround for this was to handle the db.name and eth1.contract.address specifically (because they both came from cli and/or defaults):
```Typescript
interface IConfigFile extends JsonMap{
  db?: {name: string};
  chain?: {chain: string};
  rpc?: {port: string};
  eth1?: {contract?: {address: string}};
}
...
    let parsedConfig: IConfigFile;
    if (options.configFile) {
      let data: Buffer;
      try {
        data = fs.readFileSync(options.configFile);
      } catch {
        throw new CliError(`${options.configFile} could not be parsed.`);
      }
      parsedConfig = parse(data.toString());
    }
      
    let optionsMap: BeaconNodeCtx = {
      db: {
        name: options.db || parsedConfig ? parsedConfig.db.name : defaults.db.name
      },
...
```

and deepmerge the rest after. I avoided handling each value specifically and still relying on deepmerge for the others (since all others worked as intended which leads me to believe it is because name was defined for that object but unset).  

Currently, the config file only saves the chain, db, rpc port, and eth1 contract address but it can be extended to include any others from the eth1 and p2p object defaults easily (I ignored other variables because the implementations were not final and I personally didn't see the need to have those config saved).

Other notes:
- I changed --depositContract abbreviation to -dc to allow config file flag to be -c
- Created interfaces for config options to allow the objects to be passed into the action function without destructuring or defaulting to any type
- Fixed a few bugs (can elaborate if you want specifics)
- Added types where missing
- Fixed other linter warnings in directly relating files
- Created basic test for if the command line function was added and that it throws the error when output flag is not set

### CODE REVIEW

**Review by wemeetagain — DISMISSED** [review]
:heart_eyes: looks good

**Comment by mpetrunic** [comment]
> I had an issue with the deepmerge library that was used in the project for my specific case. The problem I had was if an empty object (which had a parameter defined and defaulted to be undefined) was deep merged with an object that had that type defined, it would not be merged and instead kept the empty object value.

@Aabell8 I've already created simple helper to overcome deepmerge wierd behaviors with objects, you can use it like here: https://github.com/ChainSafe/lodestar/blob/master/src/node/index.ts#L55

File: src/cli/commands/beacon.ts [hunk 1]
```diff
@@ -1,55 +1,90 @@
...
+      } catch {
+        throw new CliError(`${options.configFile} could not be parsed.`);
+      }
+      parsedConfig = parse(data.toString());
+    }
+    
+    let optionsMap: BeaconNodeCtx = {
       db: {
-        name: options.db
+        name: options.db || parsedConfig ? parsedConfig.db.name : defaults.db.name
```
Comment (mpetrunic) [inline src/cli/commands/beacon.ts]: If you apply my deepmerge from comment you should be able to remove this :)

File: src/cli/commands/beacon.ts [hunk 2]
```diff
@@ -1,55 +1,90 @@
...
-
       });
   }
 
-  public async action(options: any): Promise<void> {
-    const node = new BeaconNode({
+  public async action(options: IBeaconCommandOptions): Promise<void> {
+    let parsedConfig: IConfigFile;
+    if (options.configFile) {
+      let data: Buffer;
```
Comment (mpetrunic) [inline src/cli/commands/beacon.ts]: I think it's going to be easier to test if you move config file loading and parsing into separate util class

File: src/cli/commands/create-config.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  public async action(options: ICreateConfigOptions): Promise<void> {
+    if (options.outputFile) {
+      if (fs.existsSync(options.outputFile)){
+        throw new CliError(`${options.outputFile} already exists`);
+      }
+
+      // Stringify defaults into TOML format
+      const defaultsToWrite = (({chain, db, rpc, eth1}) => ({
```
Comment (mpetrunic) [inline src/cli/commands/create-config.ts]: Config generation could also go in config util class. Generally these commander methods are awkward to test so we should keep them as small as possible

File: src/cli/commands/create-config.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+          contract: {
+            address: eth1.depositContract.address
+          }
+        }
+      }))(defaults);
+
+      const tomlString = stringify(defaultsToWrite);
+
+      // Save TOML formatted to output file
+      fs.writeFile(options.outputFile, tomlString, err => {
```
Comment (mpetrunic) [inline src/cli/commands/create-config.ts]: https://nodejs.org/api/fs.html#fs_fs_writefilesync_file_data_options

If you use sync methods instead of callback ones you won't get unhandled promise rejections

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Looks good! Just few small code organization comments and deepmerge fix

File: src/cli/commands/create-config.ts (hunk 4 above)
Comment (GregTheGreek) [reply src/cli/commands/create-config.ts]: +1

File: src/cli/commands/create-config.ts (hunk 3 above)
Comment (GregTheGreek) [reply src/cli/commands/create-config.ts]: could also just add a method to the `CreateConfigClass`, not sure how useful it will be to have a seperate file for.

File: src/cli/commands/beacon.ts (hunk 1 above)
Comment (austinabell) [reply src/cli/commands/beacon.ts]: will do, I had looked into the library and assumed it wouldn't solve my issue and I thought I had tested it to not fix in my case, and based on your comment I thought you had a very specific use for that config option. I will try it out now to test if it solves my issue, thanks.

File: src/cli/commands/beacon.ts (hunk 2 above)
Comment (austinabell) [reply src/cli/commands/beacon.ts]: Great suggestion, I was planning on doing this especially since the 2 code smells relating to this function. The other 3 code smells seems very wrong to fix to me, is there a `mark as won't fix` equivalent on codeclimate or do you guys think I should change that in some way.

for reference the code smell was referencing these lines:

```Typescript
commander
      .command("beacon")
      .description("Start lodestar node")
      .option("-d, --db [db_path]", "Path to file database")
      .option("-dc, --depositContract [address]", "Address of deposit contract")
```

File: src/cli/commands/create-config.ts (hunk 4 above)
Comment (austinabell) [reply src/cli/commands/create-config.ts]: gotcha, of course. I had switched the reading to synchronous and I guess I had forgot to loop back around. Thanks

File: src/cli/commands/beacon.ts (hunk 2 above)
Comment (wemeetagain) [reply src/cli/commands/beacon.ts]: I marked those code smells as wontfix

File: src/cli/commands/beacon.ts (hunk 1 above)
Comment (austinabell) [reply src/cli/commands/beacon.ts]: Yeah, I just tested again, for some reason for specifically the db object (and even when I change type to any) does not merge any values that are in the 
```Typescript
let optionsMap: BeaconNodeCtx = {
      db: {
        name: options.db,
      },
      eth1: {
        contract: {
          address: options.depositContract
        },
        provider: await this.getProvider(options.eth1RpcUrl)
      },
      rpc: {
        apis: this.setupRPC(options.rpc)
      }
};
```
object, but for any values in the toml not matching this pattern are merged. The weirdest part is the eth1.contract.address is merged correctly even though that one is defined as well. I can push a commit to make it easy to test yourself to see. I will leave just the db overriding in just to ensure it still works in the commit. To be clear this issue happens even without the config file flag addition from this commit, the only reason it was working was because a default was passed in through an option and set manually (which can't be used here because of the hierarchy of config needed).

To replicate the issue (When I push my next commit fixing other issues as well):
- remove the logic to handle db name specifically ie: `name: options.db,`
- Run the beacon command with or without config file (and without the -d flag), you will see that the db will be an empty object when BeaconNode is created

File: src/cli/commands/create-config.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+interface ICreateConfigOptions {
+  outputFile: string;
+}
+
+export class CreateConfigCommand implements CliCommand {
+  public register(commander: CommanderStatic): void {
+    commander
+      .command("create-config")
+      .description("Create default config file")
+      .option("-o, --outputFile [output_file]", "Path to output file destination")
```
Comment (GregTheGreek) [inline src/cli/commands/create-config.ts]: Should add default output file, perhaps `.option("-o, --outputFile [output_file]", "Path to output file destination", "lodestar-config.toml")`

File: src/cli/commands/create-config.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      if (fs.existsSync(options.outputFile)){
+        throw new CliError(`${options.outputFile} already exists`);
+      }
+
+      writeTomlConfig(options.outputFile);
+
+      logger.info(`Successfully wrote config file to ${options.outputFile}`);
+    } else {
+      throw new CliError("A file must be specified using the -o flag");
+    }
```
Comment (GregTheGreek) [inline src/cli/commands/create-config.ts]: based on above comment the main if statement (`if (options.outputFile) {...}`) is unecessary

File: src/cli/commands/deposit.ts [hunk 7]
```diff
@@ -18,31 +26,32 @@ export class DepositCommand implements CliCommand {
...
+      throw new CliError(`JSON RPC node (${options.node}) not available. Reason: ${e.message}`);
     }
 
     const wallets = [];
-    if(mnemonic) {
-      wallets.push(...this.fromMnemonic(mnemonic, provider, 10));
-    } else if (privateKey) {
-      wallets.push(new Wallet(privateKey, provider));
+    if(options.mnemonic) {
+      wallets.push(...this.fromMnemonic(options.mnemonic, provider, 10));
```
Comment (GregTheGreek) [inline src/cli/commands/deposit.ts]: Since you're fixing the mnemonic, we should allow an override for `n` number of wallets and default it int he flag to 10. Something like:
`option.( -a, --accounts, "Number of accounts to generate at startup", 10)`

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

File: src/cli/commands/create-config.ts (hunk 5 above)
Comment (austinabell) [reply src/cli/commands/create-config.ts]: awesome, yeah I'll do that now. Do you want it to default to a file if the -o / --outputFile flag is not set at all? IE: `./bin/lodestar create-config` would just create using a default file?

File: src/cli/commands/deposit.ts (hunk 7 above)
Comment (austinabell) [reply src/cli/commands/deposit.ts]: will do !

Comment (austinabell) [reply src/cli/commands/deposit.ts]: Add this default to defaults and save it with create-config, or is it not important?

**Comment by austinabell** [comment]
One last check, do you guys want me to add any other variables to save to this config file, or did I cover all of the ones you would expect?

File: src/util/toml.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+ * @returns {IConfigFile}
+ */
+export function getTomlConfig(fileName: string): IConfigFile {
+  let data: Buffer;
+  try {
+    data = fs.readFileSync(fileName);
+  } catch {
+    throw new CliError(`${fileName} could not be parsed.`);
+  }
+  return parse(data.toString());
```
Comment (mpetrunic) [inline src/util/toml.ts]: put this line inside try clause if it throws while parsing

**Review by mpetrunic — CHANGES_REQUESTED** [review]
one more thing and looks good to me

**Review by mpetrunic — DISMISSED** [review]

**Review by GregTheGreek — DISMISSED** [review]

**Review by wemeetagain — DISMISSED** [review]

**Comment by wemeetagain** [comment]
theres a merge conflict that has to be resolved

**Review by mpetrunic — APPROVED** [review]

**Comment by austinabell** [comment]
Sorry was pretty off grid this weekend, I resolved the conflicts and tested with the merged branch and everything seems to function as intended.

**Review by GregTheGreek — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]
